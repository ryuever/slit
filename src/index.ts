export { default as sortWithOptions } from './sortWithOptions';
