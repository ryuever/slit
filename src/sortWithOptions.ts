/**
 *
 * @param items items to process
 * @param getOptions return an array, which is used to constraint the order of on `${index}th` loop
 * @param getOptionsCount limit the recursive count
 * @param getOption get value from item, which is used for comparison..
 *
 */

export interface sortWithOptionsProps<T, K> {
  items: Array<T>;
  getOptions: (index: number) => Array<K>;
  getOption: (item: T, index: number) => K;
  getOptionsCount?: () => number;
}

export interface sortWithOptionsInternalProps<T> {
  items: Array<T>;
  index: number;
}

const sortWithOptions = <T extends object = object, K extends any = any>({
  items,
  getOptions,
  getOption,
  getOptionsCount,
}: sortWithOptionsProps<T, K>) => {
  const sorter = ({ items, index }: sortWithOptionsInternalProps<T>) => {
    const options = getOptions(index);
    items.sort((a, b) => {
      const aIndex = options.indexOf(getOption(a, index));
      const bIndex = options.indexOf(getOption(b, index));
      return aIndex - bIndex;
    });

    const result: Array<Array<T>> = [];
    let parts: Array<T> = [];

    for (let i = 0; i < items.length; i++) {
      const current = items[i];
      const last = parts[parts.length - 1];

      if (!last) {
        parts.push(current);
      } else if (getOption(current, index) === getOption(last, index)) {
        parts.push(current);
      } else {
        result.push(parts);
        parts = [current];
      }
    }

    if (parts.length) {
      result.push(parts);
      parts = [];
    }

    const nextIndex = index + 1;
    let next: Array<Array<T>> = [];
    let shouldRecursiveCheck = false;

    if (typeof getOptionsCount === 'function') {
      const count = getOptionsCount();
      if (nextIndex < count) shouldRecursiveCheck = true;
    } else {
      try {
        if (getOptions(nextIndex)) shouldRecursiveCheck = true;
      } catch (err) {
        shouldRecursiveCheck = false;
      }
    }

    if (shouldRecursiveCheck) {
      result.forEach(group => {
        next.push(
          sorter({
            items: group,
            index: nextIndex,
          })
        );
      });
    } else {
      next = result;
    }

    return next.reduce((values, cur) => {
      return values.concat(cur);
    }, []);
  };

  return sorter({
    items,
    index: 0,
  });
};

export default sortWithOptions;
