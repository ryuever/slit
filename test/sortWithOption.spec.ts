import { sortWithOptions } from '../src';

describe('Two columns', () => {
  const items = [
    {
      slugs: ['丝绒款09 #草莓粉', '4.8g'],
    },
    {
      slugs: ['丝绒款10 #珊瑚红', '5.2g'],
    },
    {
      slugs: ['丝绒款10 #珊瑚红', '3.0g'],
    },
    {
      slugs: ['丝绒款10 #珊瑚红', '7.0g'],
    },
    {
      slugs: ['丝绒款10 #珊瑚红', '4.8g'],
    },
    {
      slugs: ['果汁款03 #西瓜红', '3.0g'],
    },
    {
      slugs: ['果汁款09 #荔枝色', '4.8g'],
    },
    {
      slugs: ['丝绒款10 #珊瑚红', '3.0g'],
    },
  ];

  const skuOptions = [
    {
      id: '5a60c42f69bd891ed8939a3c',
      name: '颜色/款式',
      values: [
        '丝绒款01 #桃红色',
        '丝绒款03 #成熟红',
        '丝绒款06 #暗红棕',
        '丝绒款07 #雾面正红',
        '丝绒款09 #草莓粉',
        '丝绒款10 #珊瑚红',
        '丝绒款11 #橘暖红',
        '果汁款02 #红柚色',
        '果汁款03 #西瓜红',
        '果汁款07 #红枣色',
        '果汁款08 #柿子色',
        '果汁款09 #荔枝色',
      ],
      soldOutValues: [
        '丝绒款06 #暗红棕',
        '丝绒款07 #雾面正红',
        '丝绒款10 #珊瑚红',
        '果汁款07 #红枣色',
      ],
    },
    {
      id: '5a60c42f69bd891ed8939a40',
      name: '重量',
      values: ['3.0g', '4.8g', '5.2g', '7.0g'],
      soldOutValues: [],
    },
  ];

  test('basic works', () => {
    const result = sortWithOptions({
      items,
      getOptions: index => {
        return skuOptions[index].values;
      },
      getOption: (item, index) => {
        return item.slugs[index];
      },
      getOptionsCount: () => skuOptions.length,
    });
    expect(result).toMatchSnapshot();
  });
});

describe('Four columns', () => {
  const items = [
    {
      slugs: ['姜黄色', 'aa', '3'],
    },
    {
      slugs: ['姜黄色', 'dd', '1'],
    },
    {
      slugs: ['姜黄色', 'cc', '5'],
    },
    {
      slugs: ['姜黄色', 'aaaad', '5', 'LL'],
    },
    {
      slugs: ['姜黄色', 'cc', '1', 'L'],
    },
    {
      slugs: ['姜黄色', 'ee', '4'],
    },
    {
      slugs: ['姜黄色', 'cc', '3', 'M'],
    },
    {
      slugs: ['姜黄色', 'dd', '2', 'XS'],
    },
    {
      slugs: ['姜黄色', 'aa', '1', 'LL'],
    },
    {
      slugs: ['姜黄色', 'dd', '1', 'XS'],
    },
    {
      slugs: ['姜黄色', 'cc', '1', 'M'],
    },
    {
      slugs: ['姜黄色', 'bb', '1', 'L'],
    },
    {
      slugs: ['姜黄色', 'cc', '5', 'XS'],
    },
    {
      slugs: ['姜黄色', 'cc', '5', 'LL'],
    },
  ];

  const skuOptions = [
    {
      id: '5a60c43a69bd891ed893aaf6',
      name: '颜色',
      values: ['姜黄色', '姜色', '湖蓝色', '粉红色', '紫色'],
      soldOutValues: [],
    },
    {
      id: '5a60c42f69bd891ed8939a46',
      name: '套装明细',
      values: ['aa', 'aaaad', 'bb', 'cc', 'dd', 'ee'],
      soldOutValues: [],
    },
    {
      id: '',
      name: '组合数量',
      values: ['1', '2', '3', '4', '5'],
      soldOutValues: [],
    },
    {
      id: '',
      name: '尺寸',
      values: ['XS', 'S', 'M', 'L', 'LL'],
      soldOutValues: [],
    },
  ];

  test('Four columns', () => {
    const result = sortWithOptions({
      items,
      getOptions: index => {
        return skuOptions[index].values;
      },
      getOption: (item, index) => {
        return item.slugs[index];
      },
      getOptionsCount: () => skuOptions.length,
    });
    expect(result).toMatchSnapshot();
  });
});
