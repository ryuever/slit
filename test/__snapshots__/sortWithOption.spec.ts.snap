// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Four columns Four columns 1`] = `
Array [
  Object {
    "slugs": Array [
      "姜黄色",
      "aa",
      "1",
      "LL",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "aa",
      "3",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "aaaad",
      "5",
      "LL",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "bb",
      "1",
      "L",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "1",
      "M",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "1",
      "L",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "3",
      "M",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "5",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "5",
      "XS",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "cc",
      "5",
      "LL",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "dd",
      "1",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "dd",
      "1",
      "XS",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "dd",
      "2",
      "XS",
    ],
  },
  Object {
    "slugs": Array [
      "姜黄色",
      "ee",
      "4",
    ],
  },
]
`;

exports[`Two columns basic works 1`] = `
Array [
  Object {
    "slugs": Array [
      "丝绒款09 #草莓粉",
      undefined,
      "4.8g",
    ],
  },
  Object {
    "slugs": Array [
      "丝绒款10 #珊瑚红",
      "3.0g",
    ],
  },
  Object {
    "slugs": Array [
      "丝绒款10 #珊瑚红",
      "3.0g",
    ],
  },
  Object {
    "slugs": Array [
      "丝绒款10 #珊瑚红",
      "4.8g",
    ],
  },
  Object {
    "slugs": Array [
      "丝绒款10 #珊瑚红",
      "5.2g",
    ],
  },
  Object {
    "slugs": Array [
      "丝绒款10 #珊瑚红",
      "7.0g",
    ],
  },
  Object {
    "slugs": Array [
      "果汁款03 #西瓜红",
      "3.0g",
    ],
  },
  Object {
    "slugs": Array [
      "果汁款09 #荔枝色",
      "4.8g",
    ],
  },
]
`;
